enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+ISO and boot hosts.
+
+A new field in the ClusterDeployment spec will optionally reference an
+AgentDrivenClusterInstallation resource, which describes everything that the
+agent-based install tooling needs to know about the cluster that it will be
+creating. Some or all of that new resource may be folded into ClusterDeployment
+in the future, but in the immediate term, keeping it separate will enable much
+quicker development.
+
+Hive or ACM will optionally orchestrate the booting of bare metal hosts by:
+
+1. Get an ISO URL from assisted operator.
+1. Use metal3's BareMetalHost to boot the live ISO on some hardware.
+
+#### BareMetalHost can boot live ISOs
+
+A [separate enhancement to
+metal3](https://github.com/metal3-io/metal3-docs/pull/150) proposes a new
+capability in the BareMetalHost API enabling it to boot live ISOs. That feature
+is required so that automation in a cluster can boot the discovery ISO on known
+hardware as the first step toward provisioning that hardware.
+
+#### CAPBM 
+
+[Cluster-API Provider Bare Metal](https://github.com/openshift/cluster-api-provider-baremetal/) will need to gain the ability to interact with the new assisted installer APIs. Specifically it will need to match a Machine with an available assisted Agent that is ready to be provisioned. This will be in addition to matching a Machine with a BareMetalHost, which it already does today.
+
+#### Day 2 Add Node Boot-it-Yourself
+
+This scenario takes place within a stand-alone bare metal OpenShift cluster.
+
+1. The user downloads a discovery ISO from the cluster. The download is implemented by the assisted installer as a URL on a AssistedInstallCluster
+resource.
+1. A host boots the live ISO, the assisted agent starts running, and the agent contacts the assisted service to register its existence. Communication utilizes the existing non-k8s REST API. The agent walks through the validation and inspection workflow as it exists today.
+1. The wrapping controller creates a new AssistedInstallAgent resource to be the k8s-native API for the agent.
+1. CAPBM creates a new BareMetalHost resource, setting the status annotation based on inspection information from the prior step.
+1. The user approves the host for installation. (TODO: how?)
+1. The user or an orchestrator scales up a MachineSet, causing a new Machine resource to be created.
+1. CAPBM binds the Machine to the BareMetalHost, as it does today. It additionally finds the CR representing the agent and uses it to begin installation.
+1. The assisted service initiates installation of the host.
+1. CAPBM updates the status on the BareMetalHost to reflect that it has been provisioned.
+
+#### Day 2 Add Node Virtualmedia Stand-alone
+
+This scenario takes place within a stand-alone bare metal OpenShift cluster.
+
+1. The user creates a BareMetalHost that includes BMC credentials and a label indicating it should be used with assisted installer.
```

Comment:
Don't we usually use annotations for API-like behaviors like this? Is the label for a labelSelector in a MachineSet?

Reply:
The idea here is that something can watch BMHs with a label selector and add the appropriate ISO URL. There are multiple ways this could be implemented though, but since label filtering is a feature of the Watch operation, it seems like a reasonable starting point.

The label would not correspond to a MachineSet.